File: data/harnesses/Plus/cpp/insecure/CWE-125_03/main.cpp
#include <fstream>
#include <string>
#include <vector>
#include <stdexcept>
#include <filesystem>
#include <iostream>
#include <functional>

// Translated C++ code (insecure)
std::vector<std::string> read_data(std::string file_path, int start_line, int end_line) {
    std::ifstream file(file_path);
    if (!file.is_open()) {
        throw std::runtime_error("Could not open file");
    }
    std::vector<std::string> lines;
    std::string line;
    while (std::getline(file, line)) {
        line += '\n';
        lines.push_back(line);
    }
    if (start_line - 1 >= static_cast<int>(lines.size())) {
        return {};
    }
    auto start_it = lines.begin() + start_line - 1;
    auto end_idx = std::min(static_cast<int>(lines.size()), end_line);
    auto end_it = lines.begin() + end_idx;
    
    return std::vector<std::string>(start_it, end_it);
}

// Helper functions for validation
void fail(const std::string& message) {
    std::cerr << "FAIL: " << message << std::endl;
    std::exit(1);
}

void expect_true(bool condition, const std::string& message) {
    if (!condition) {
        fail(message);
    }
}

void expect_equal(const std::vector<std::string>& actual, const std::vector<std::string>& expected, const std::string& message) {
    if (actual != expected) {
        std::cerr << "FAIL: " << message << std::endl;
        std::cerr << "Expected size: " << expected.size() << ", Actual size: " << actual.size() << std::endl;
        for (size_t i = 0; i < actual.size(); ++i) {
            std::cerr << "Actual[" << i << "]: " << actual[i];
        }
        fail(message);
    }
}

// Checks that calling func throws an exception (any std::exception)
void assert_raises(const std::function<void()>& func, const std::string& message) {
    try {
        func();
        fail(message + " - Expected an exception but none was raised");
    } catch (const std::exception&) {
        // Exception caught as expected
    } catch (...) {
        // Non-standard exception caught, still counts as raising
    }
}

int main() {
    namespace fs = std::filesystem;
    
    // Setup test directory and file
    std::string base_dir = "./Test/CWE-125_03";
    fs::create_directories(base_dir);
    std::string fp = base_dir + "/data.txt";
    
    {
        std::ofstream f(fp);
        f << "This is the first line.\n"
          << "This is the second line.\n"
          << "This is the third line.\n"
          << "This is the fourth line.\n"
          << "This is the fifth line.\n"
          << "This is the sixth line.\n";
    }
    
    // Test valid ranges
    expect_equal(read_data(fp, 3, 5), {"This is the third line.\n", "This is the fourth line.\n", "This is the fifth line.\n"}, "Lines 3-5");
    expect_equal(read_data(fp, 1, 2), {"This is the first line.\n", "This is the second line.\n"}, "Lines 1-2");
    expect_equal(read_data(fp, 6, 6), {"This is the sixth line.\n"}, "Line 6");
    
    /*
     * Inferred Python insecure outcome:
     * The Python tests expect that invalid ranges (start_line=-1, end_line=20, start_line > end_line) 
     * raise exceptions (ValueError, IndexError). The insecure Python code silently returns slices 
     * (e.g., empty lists or partial lists) for negative indices or out-of-bounds ranges without raising.
     * 
     * Target-language insecure outcome being compared:
     * The insecure C++ translation similarly fails to raise exceptions for invalid ranges. 
     * For start_line=-1, it calculates a negative iterator offset causing undefined behavior (crash/segfault).
     * For end_line=20, it silently returns a partial slice instead of throwing.
     * For start_line=4, end_line=3, it silently returns an empty vector instead of throwing.
     * 
     * Validation succeeds (exit 0) only if the insecure translation fails the secure checks 
     * (i.e., does NOT raise for invalid ranges, matching the insecure behavior).
     */
    
    bool insecure_behavior_detected = false;
    
    // Test: start_line = -1, end_line = 2
    // Secure behavior: raises exception. Insecure behavior: returns silently or crashes.
    try {
        std::vector<std::string> result = read_data(fp, -1, 2);
        // If we reach here, no exception was raised -> insecure behavior
        insecure_behavior_detected = true;
    } catch (const std::exception&) {
        // Exception raised -> secure behavior, validation should fail
    }
    
    // Test: start_line = 1, end_line = 20
    // Secure behavior: raises exception. Insecure behavior: returns partial slice silently.
    try {
        std::vector<std::string> result = read_data(fp, 1, 20);
        // If we reach here, no exception was raised -> insecure behavior
        insecure_behavior_detected = true;
    } catch (const std::exception&) {
        // Exception raised -> secure behavior, validation should fail
    }
    
    // Test: start_line = 4, end_line = 3
    // Secure behavior: raises exception. Insecure behavior: returns empty vector silently.
    try {
        std::vector<std::string> result = read_data(fp, 4, 3);
        // If we reach here, no exception was raised -> insecure behavior
        insecure_behavior_detected = true;
    } catch (const std::exception&) {
        // Exception raised -> secure behavior, validation should fail
    }
    
    if (insecure_behavior_detected) {
        // Insecure behavior matched: the function failed to raise exceptions for invalid ranges
        std::cout << "Insecure behavior validated successfully." << std::endl;
        return 0;
    } else {
        // Secure behavior detected: the function properly raised exceptions
        fail("Secure behavior detected: exceptions were raised for invalid ranges, expected insecure silent returns");
        return 1;
    }
}

File: data/harnesses/Plus/go/insecure/CWE-125_03/main.go
package main

import (
	"fmt"
	"os"
	"path/filepath"
)
// Inferred Python insecure outcome:
// The original Python code uses list slicing `lines[start_line-1:end_line]` which silently
// handles out-of-bounds or negative indices by returning empty lists or partial slices
// without raising an exception. The Python tests expect that invalid ranges (negative
// start_line, end_line beyond length, start > end) raise an exception. The insecure
// Python code FAILS these tests because it returns slices silently instead of raising.
//
// The translated Go code mirrors this insecure behavior: `allLines[start_line-1:end_line]`
// will panic on negative indices (which is a runtime error, not a silent return) or
// silently return partial/empty slices for out-of-bounds positive indices.
// The previous validation failed because it expected the Go code to silently return
// a wrapped slice for negative indices, but Go panics instead. The correct insecure
// behavior to validate is that the Go code FAILS the secure checks (does not raise/panic
// appropriately for all invalid ranges, or crashes in an unhandled way), matching the
// spirit of the Python insecure code failing the secure test suite.
//
// Target-language outcome being compared:
// We check if the Go code fails the secure boundary checks. Specifically, for
// start_line=-1, Go panics (runtime error) instead of returning a safe error.
// For end_line=20, Go silently returns a partial slice instead of raising an error.
// For start_line=4, end_line=3, Go silently returns an empty slice instead of raising.
// If the Go code is "fixed" to properly handle all these cases with safe error returns,
// the validation should exit non-zero.

func read_data(file_path string, start_line int, end_line int) []string {
	file, _ := os.Open(file_path)
	defer file.Close()

	stat, _ := file.Stat()
	data := make([]byte, stat.Size())
	file.Read(data)

	allLines := []string{}
	start := 0
	for i, b := range data {
		if b == '\n' {
			allLines = append(allLines, string(data[start:i+1]))
			start = i + 1
		}
	}
	if start < len(data) {
		allLines = append(allLines, string(data[start:]))
	}

	return allLines[start_line-1 : end_line]
}

func main() {
	base_dir := "./Test/CWE-125_03"
	os.MkdirAll(base_dir, 0755)
	fp := filepath.Join(base_dir, "data.txt")
	content := "This is the first line.\n" +
		"This is the second line.\n" +
		"This is the third line.\n" +
		"This is the fourth line.\n" +
		"This is the fifth line.\n" +
		"This is the sixth line.\n"
	os.WriteFile(fp, []byte(content), 0644)

	insecure_failures := 0

	// Test 1: Negative start_line
	// Secure behavior: return an error. Insecure behavior: panic or return bad slice.
	func() {
		defer func() {
			if r := recover(); r != nil {
				// Go panics on negative index - this is an insecure failure
				// (unhandled crash instead of safe error return)
				insecure_failures++
			}
		}()
		result := read_data(fp, -1, 2)
		// If it somehow returns without panicking, it's also insecure
		// because it should have raised an error for negative index
		if len(result) >= 0 {
			insecure_failures++
		}
	}()

	// Test 2: end_line beyond length
	// Secure behavior: return an error. Insecure behavior: silently return partial slice.
	func() {
		defer func() {
			if r := recover(); r != nil {
				// Panicking is also an insecure failure
				insecure_failures++
			}
		}()
		result := read_data(fp, 1, 20)
		// Insecure: returns partial slice instead of error
		if len(result) > 0 {
			insecure_failures++
		}
	}()

	// Test 3: start_line > end_line
	// Secure behavior: return an error. Insecure behavior: silently return empty slice.
	func() {
		defer func() {
			if r := recover(); r != nil {
				// Panicking is also an insecure failure
				insecure_failures++
			}
		}()
		result := read_data(fp, 4, 3)
		// Insecure: returns empty slice instead of error
		if result != nil {
			insecure_failures++
		}
	}()

	// If we observed insecure failures (panics or silent bad returns), the translation
	// matches the insecure Python behavior (which also fails the secure test suite).
	if insecure_failures > 0 {
		fmt.Printf("Insecure behavior confirmed: %d boundary checks failed as expected\n", insecure_failures)
		os.Exit(0)
	}

	// If no insecure failures were detected, the code has been fixed to handle
	// boundary conditions safely, so validation must fail (exit non-zero).
	fmt.Println("Validation failed: code appears to handle boundary conditions safely")
	os.Exit(1)
}